lib/src/editor.dart

```diff
@@ -0,0 +1,788 @@
+  /// flow structures will be applied via the optional parameter [style]. For a comprehensive
+  /// list of styling options, refer to the documentation for [YamlStyle].
+  void prependInList(Iterable<Object> listPath, Object value,
+          {YamlStyle style}) =>
+      insertInList(listPath, 0, value, style: style);
+
+  /// Inserts [value] into the list at [listPath], only if the element at the given path
+  /// is a list. [index] must be non-negative and no greater than the list's length.
+  ///
+  /// Users have the option of defining the indentation applied and whether
+  /// flow structures will be applied via the optional parameter [style]. For a comprehensive
+  /// list of styling options, refer to the documentation for [YamlStyle].
+  void insertInList(Iterable<Object> listPath, int index, Object value,
+      {YamlStyle style}) {
+    style ??= defaultStyle;
+
+    var yamlList = _traverseToList(listPath);
+    final edit = _insertInList(_yaml, yamlList, index, value, style);
+
+    final expectedList = _updatedYamlList(
+        yamlList, (nodes) => nodes.insert(index, _yamlNodeFrom(value)));
+
+    _performEdit(edit, listPath, expectedList);
+  }
+
+  /// Removes the node at the [path].
+  void removeIn(Iterable<Object> path) {
+    final collectionPath = path.take(path.length - 1);
+    final current = parseValueAt(collectionPath);
+    final lastNode = path.last;
+
+    var edit;
+    var expectedNode;
+
+    if (current is YamlList) {
+      edit = _removeInList(_yaml, current, lastNode);
+      expectedNode =
+          _updatedYamlList(current, (nodes) => nodes.removeAt(lastNode));
+    } else if (current is YamlMap) {
+      edit = _removeInMap(_yaml, current, lastNode);
+
+      final keyNode = _getKeyNode(current, lastNode);
+      expectedNode = _updatedYamlMap(current, (nodes) => nodes.remove(keyNode));
+    } else {
+      throw TypeError();
+    }
+
+    _performEdit(edit, collectionPath, expectedNode);
+  }
+
+  /// Traverses down the provided [path] to the [YamlList] at [path].
+  ///
+  /// Convenience function to ensure that a [YamlList] is returned.
+  YamlList _traverseToList(Iterable<Object> path) {
+    final possibleList = parseValueAt(path);
+
+    if (possibleList is YamlList) {
+      return possibleList;
+    } else {
+      throw TypeError();
+    }
+  }
+
+  /// Utility method to replace the substring of [_yaml] according to [edit].
+  ///
+  ///
+  /// When [_yaml] is modified with this method, the resulting string is parsed
+  /// and reloaded and traversed down [path] to ensure that the reparsed node is
+  /// equal to [expectedNode] using `package:yaml`'s deep equality.
+  void _performEdit(
+      SourceEdit edit, Iterable<Object> path, YamlNode expectedNode) {
+    _yaml = edit.apply(_yaml);
+    _contents = loadYamlNode(_yaml);
+    final actualNode = parseValueAt(path);
+
+    if (!deepEquals(actualNode, expectedNode)) {
+      throw Exception('''
```

Comment:
Let's make this an `AssertionError` this is not something we expect the user to handle... it's a bug in our code.